### PR TITLE
consensus: fix_step_votes_registry

### PR DESCRIPTION
**Pull request by herr-seppia**
Fix missing collection of failed cert when a valid vote is received firstly

See also #1255

### CODE REVIEW

File: consensus/src/step_votes_reg.rs [hunk 1]
```diff
@@ -94,33 +83,38 @@ impl CertificateInfo {
...
+            && self.quorum_reached_ratification
+            && self.quorum_reached_validation
     }
 }
 
 pub type SafeCertificateInfoRegistry = Arc<Mutex<CertInfoRegistry>>;
 
+#[derive(Default, Clone)]
+struct CertInfo {
+    certs: HashMap<[u8; 32], CertificateInfo>,
```
Comment (goshawk-3) [inline consensus/src/step_votes_reg.rs:102]: If the case is to allow only two options - Nil and Valid CertificateInfo, HashMap is a bit overkill here

File: consensus/src/step_votes_reg.rs [hunk 2]
```diff
@@ -49,19 +47,10 @@ impl CertificateInfo {
     pub(crate) fn add_sv(
         &mut self,
         iter: u8,
-        hash: [u8; 32],
         sv: StepVotes,
         svt: SvType,
         quorum_reached: bool,
     ) -> bool {
-        if let Some(h) = self.hash {
```
Comment (goshawk-3) [inline consensus/src/step_votes_reg.rs:57]: As stated, reaching a quorum in non-Emergency mode depends on the `aggregator` state but not `sv_registry` state , so I don't expect that the issue will be resolved with this change.

PS. However, I do agree the patch beneficial for failed_certificates correctness.

**Review by goshawk-3 — CHANGES_REQUESTED** [review]

File: consensus/src/step_votes_reg.rs (hunk 1 above)
Comment (herr-seppia) [reply consensus/src/step_votes_reg.rs:102]: Agree with you

However, are we sure that no other hash can be filled there? I mean, if a provisioner send a third hash, is that message discarded before right?

File: consensus/src/step_votes_reg.rs (hunk 2 above)
Comment (herr-seppia) [reply consensus/src/step_votes_reg.rs:57]: Then I could remove the "resolve" for the issue. But keep in mind that with this PR I'm not able to replicate the issue anymore, so in someway this resolves that

File: consensus/src/step_votes_reg.rs (hunk 1 above)
Comment (fed-franz) [reply consensus/src/step_votes_reg.rs:102]: A third hash means either two candidates from the same generator, or there's a fork that reached the same height (and this should be handled separately)

File: consensus/src/step_votes_reg.rs (hunk 2 above)
Comment (fed-franz) [reply consensus/src/step_votes_reg.rs:57]: > Then I could remove the "resolve" for the issue. But keep in mind that with this PR I'm not able to replicate the issue anymore, so in someway this resolves that

we should figure out WHY this PR solves that issue...

Comment (goshawk-3) [reply consensus/src/step_votes_reg.rs:57]: It's up to you. I'd rather investigate why and what this is resolving. If the problem is due to a race condition, you might be unable to reproduce it consistently.

File: consensus/src/step_votes_reg.rs [hunk 3]
```diff
@@ -94,33 +83,58 @@ impl CertificateInfo {
     fn is_ready(&self) -> bool {
         !self.cert.validation.is_empty()
             && !self.cert.ratification.is_empty()
-            && self.hash.is_some()
             && self.quorum_reached_validation
             && self.quorum_reached_ratification
     }
 
     /// Returns `true` if the certificate has empty hash
```
Comment (fed-franz) [inline consensus/src/step_votes_reg.rs:90]: /// Returns `true` if  the certificate is for a NilQuorum

Comment (fed-franz) [reply consensus/src/step_votes_reg.rs:90]: Are you excluding partial (non-quorum) certificates?

File: consensus/src/step_votes_reg.rs [hunk 4]
```diff
@@ -94,33 +83,58 @@ impl CertificateInfo {
...
+                None
+            }
+        }
+    }
+}
+
 pub struct CertInfoRegistry {
     ru: RoundUpdate,
 
     /// List of iterations certificates. Position in the array represents
```
Comment (fed-franz) [inline consensus/src/step_votes_reg.rs:128]: Needs to be updated

File: consensus/src/step_votes_reg.rs [hunk 5]
```diff
@@ -22,7 +22,7 @@ pub(crate) enum SvType {
 #[derive(Default, Clone, Copy)]
 struct CertificateInfo {
     /// represents candidate block hash
```
Comment (fed-franz) [inline consensus/src/step_votes_reg.rs:24]:
```suggestion
    /// represents vote (candidate hash or nil)
```

**Review by fed-franz — CHANGES_REQUESTED** [review]
few nits

File: consensus/src/step_votes_reg.rs (hunk 3 above)
Comment (fed-franz) [inline consensus/src/step_votes_reg.rs:90]:
```suggestion
    /// Returns `true` if the certificate is a NilQuorum
```

File: consensus/src/step_votes_reg.rs [hunk 6]
```diff
@@ -177,18 +190,18 @@ impl CertInfoRegistry {
...
-                res.push(None)
-            }
+        let mut res = Vec::with_capacity(to as usize);
+
+        for iteration in 0u8..to {
+            res.push(
+                self.cert_list
+                    .get(&iteration)
+                    .map(|c| c.nil)
+                    .filter(|ci| ci.is_nil_quorum())
```
Comment (fed-franz) [inline consensus/src/step_votes_reg.rs:202]: Are you excluding partial (non-quorum) certificates from FailedIterations?

Comment (herr-seppia) [reply consensus/src/step_votes_reg.rs:202]: Yes, I didn't change it.

Ps: you were right then

**Review by fed-franz — APPROVED** [review]

**Review by goshawk-3 — APPROVED** [review]

**Pull request merged**